mbrl/planning/trajectory_opt.py

```diff
@@ -153,6 +153,327 @@ def optimize(
+        upper_bound: Sequence[Sequence[float]],
+        keep_elite_frac: float,
+        alpha: float,
+        device: torch.device,
+        return_mean_elites: bool = False,
+        round_population: bool = False,
+        ensemble_size: int = None,
+    ):
+        super().__init__()
+        self.num_iterations = num_iterations
+        self.elite_ratio = elite_ratio
+        self.population_size = population_size
+        self.population_decay_factor = population_decay_factor
+        self.elite_num = np.ceil(self.population_size * self.elite_ratio).astype(
+            np.int32
+        )
+        self.colored_noise_exponent = colored_noise_exponent
+        self.lower_bound = torch.tensor(lower_bound, device=device, dtype=torch.float32)
+        self.upper_bound = torch.tensor(upper_bound, device=device, dtype=torch.float32)
+        self.initial_var = ((self.upper_bound - self.lower_bound) ** 2) / 16
+        self.keep_elite_frac = keep_elite_frac
+        self.keep_elite_size = np.ceil(keep_elite_frac * self.elite_num).astype(
+            np.int32
+        )
+        self.elite = None
+        self.alpha = alpha
+        self.return_mean_elites = return_mean_elites
+        self.round_population = round_population
+        self.ensemble_size = ensemble_size
+        self.device = device
+
+    def optimize(
+        self,
+        obj_fun: Callable[[torch.Tensor], torch.Tensor],
+        x0: Optional[torch.Tensor] = None,
+        callback: Optional[Callable[[torch.Tensor, torch.Tensor, int], None]] = None,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Runs the optimization using iCEM.
+
+        Args:
+            obj_fun (callable(tensor) -> tensor): objective function to maximize.
+            x0 (tensor, optional): initial mean for the population. Must
+                be consistent with lower/upper bounds.
+            callback (callable(tensor, tensor, int) -> any, optional): if given, this
+                function will be called after every iteration, passing it as input the full
+                population tensor, its corresponding objective function values, and
+                the index of the current iteration. This can be used for logging and plotting
+                purposes.
+        Returns:
+            (torch.Tensor): the best solution found.
+        """
+        mu = x0.clone()
+        var = self.initial_var.clone()
+
+        best_solution = torch.empty_like(mu)
+        best_value = -np.inf
+
+        for i in range(self.num_iterations):
+            decay_population_size = np.ceil(
+                np.max(
+                    (
+                        self.population_size * self.population_decay_factor ** -i,
+                        2 * self.elite_num,
+                    )
+                )
+            ).astype(np.int32)
+
+            # to ensure population size is a multiple of the number of models in the ensemble
+            if self.round_population is True:
```

Comment:
I haven't looked at this carefully, but my guess is that this logic can probably be simplified. I'll try to make a more concrete suggestion later.